nucliadb/nucliadb/purge/orphan_shards.py

```diff
@@ -0,0 +1,252 @@
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import argparse
+import asyncio
+from dataclasses import dataclass
+from datetime import datetime
+
+import pkg_resources
+from grpc.aio import AioRpcError  # type: ignore
+
+from nucliadb.common.cluster import manager
+from nucliadb.common.cluster.exceptions import ShardsNotFound
+from nucliadb.common.cluster.manager import KBShardManager
+from nucliadb.common.cluster.utils import setup_cluster, teardown_cluster
+from nucliadb.common.datamanagers.rollover import RolloverDataManager
+from nucliadb.common.maindb.driver import Driver
+from nucliadb.common.maindb.utils import setup_driver, teardown_driver
+from nucliadb.ingest import logger
+from nucliadb.ingest.orm.knowledgebox import KnowledgeBox
+from nucliadb_telemetry import errors
+from nucliadb_telemetry.logs import setup_logging
+
+
+@dataclass
+class ShardLocation:
+    kbid: str
+    node_id: str
+
+
+UNKNOWN_KB = "unknown"
+
+
+async def detect_orphan_shards(driver: Driver) -> dict[str, ShardLocation]:
+    """Detect orphan shards in the system. An orphan shard is one indexed but
+    not referenced for any stored KB.
+
+    """
+    # To avoid detecting a new shard as orphan, query the index first and maindb
+    # afterwards
+    indexed_shards = await _get_indexed_shards()
+    stored_shards = await _get_stored_shards(driver)
+
+    # Log an error in case we found a shard stored but not indexed, this should
+    # never happen as shards are created in the index node and then stored in
+    # maindb
+    not_indexed_shards = stored_shards.keys() - indexed_shards.keys()
+    for shard_id in not_indexed_shards:
+        location = stored_shards[shard_id]
+        logger.error(
+            "Found a shard on maindb not indexed in the index nodes",
+            extra={
+                "shard_id": shard_id,
+                "kbid": location.kbid,
+                "node_id": location.node_id,
+            },
+        )
+
+    orphan_shard_ids = indexed_shards.keys() - stored_shards.keys()
+    orphan_shards: dict[str, ShardLocation] = {}
+    unavailable_nodes = set()
+    rollover_dm = RolloverDataManager(driver)
+    for shard_id in orphan_shard_ids:
+        node_id = indexed_shards[shard_id].node_id
+        node = manager.get_index_node(node_id)
+        kbid = UNKNOWN_KB
+
+        if node is None:
+            unavailable_nodes.add(node_id)
```

Comment:
General style/aesthetic nit comment that feel free to ignore:

If you use `break` or `continue` on while/for loops the code is much easier to understand, as there is less nested branching.

Reply:
I'm not sure I'm following you but this if is meant to skip the gRPC call when node is not available but still add this shard as orphan. Maybe I can do something to write it more clear